src/transformers/modeling_bart.py

```diff
@@ -935,12 +943,269 @@ def forward(
+
+        # scores for each sentence in the beam
+        beam_scores = torch.zeros((batch_size, num_beams), dtype=torch.float, device=input_ids.device)
+        beam_scores[:, 1:] = -1e9  # avoid ties in first step
+        beam_scores = beam_scores.view(-1)  # shape (batch_size * num_beams,)
+
+        # decoder tokens
+        prev_output_tokens = input_ids.new(batch_size * num_beams, 1).long().fill_(-1)
+        prev_output_tokens[:, 0] = 2  # HARDCODED EOS, which will be removed at the end.
+        decoder_cache = None
+        done = [False for _ in range(batch_size)]  # done sentences
+
+        self.model.decoder.generation_mode = True  # tells decoder not to use causal mask
+        for step in range(max_length + 1):
+            decoder_input_ids = prev_output_tokens.clone()
+            model_inputs = self.prepare_inputs_for_generation(
+                input_ids, decoder_cache, decoder_input_ids, attention_mask,
+            )
+            outputs = self(**model_inputs)
+            lprobs = F.log_softmax(outputs[0][:, -1, :], dim=-1)
+
+            assert pad_token_id is not None, "should break circleci"
+            lprobs[lprobs != lprobs] = -math.inf  # kill nans
+            lprobs[:, pad_token_id] = -math.inf
+            # TODO(SS): unk is 3? fairseq also takes out every step, and has unk at slot 3
+
+            assert bos_token_id == 0, "configurable bos_token_id not yet supported"
+            if step == max_length:  # FORCE EOS to be chosen
+                lprobs[:, :eos] = -math.inf
+                lprobs[:, eos + 1 :] = -math.inf
+            elif step == 0:  # Force BOS to be chosen
+                lprobs[:, bos_token_id + 1 :] = -math.inf
+            if step < min_len:
+                lprobs[:, eos] = -math.inf
+            assert self._do_output_past(outputs)
+            decoder_cache = outputs[1]
+            if repetition_penalty != 1.0:
+                self.enforce_repetition_penalty_(lprobs, batch_size, num_beams, prev_output_tokens, repetition_penalty)
+            num_hypos = batch_size * num_beams
+            if no_repeat_ngram_size > 0:  # copied from fairseq
+                # for each beam and batch sentence, generate a list of previous ngrams
+                banned_tokens = self.calc_banned_tokens(prev_output_tokens, num_hypos, no_repeat_ngram_size, step)
+                # then set their probabilities to -inf
+                for bbsz_idx in range(num_hypos):
+                    lprobs[bbsz_idx, banned_tokens[bbsz_idx]] = -math.inf
+            assert lprobs.size() == (batch_size * num_beams, vocab_size)
+            _scores = lprobs + beam_scores[:, None].expand_as(lprobs)  # (batch_size * num_beams, vocab_size)
+
+            # re-organize to group the beam together (we are keeping top hypothesis across beams)
+            _scores = _scores.view(batch_size, num_beams * vocab_size)  # (batch_size, num_beams * vocab_size)
+            next_scores, next_words = torch.topk(_scores, 2 * num_beams)
```

Comment:
should add a comment why we sample `2 * num_beams` words. In the fairseq code they do this because 
https://github.com/pytorch/fairseq/blob/fba10af9db5edd61f78ccdb9d115c4eafbcc561d/fairseq/search.py#L78

> # Take the best 2 x beam_size predictions. We'll choose the first
> # beam_size of these which don't predict eos to continue with.